File: Intermediate/29.Find the minimum number of coins needed for a given amount using a greedy approach/Solution.py
def min_coins(coins, amount):
    # Sort coins in descending order
    coins.sort(reverse=True)
    count = 0
    used_coins = []
    
    for coin in coins:
        while amount >= coin:
            amount -= coin
            count += 1
            used_coins.append(coin)
    
    return count, used_coins, amount

if __name__ == "__main__":
    n = int(input("Enter the number of coin denominations: "))
    coins = []
    print("Enter the coin denominations:")
    for _ in range(n):
        coins.append(int(input()))
    
    amount = int(input("Enter the amount: "))
    count, used_coins, remainder = min_coins(coins, amount)
    
    if remainder != 0:
        print("Exact amount cannot be formed with the given denominations.")
    else:
        print("Coins used:", used_coins)
        print("Minimum number of coins needed:", count)

File: Intermediate/29.Find the minimum number of coins needed for a given amount using a greedy approach/Solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int n;
    cout << "Enter the number of coin denominations: ";
    cin >> n;
    
    vector<int> coins(n);
    cout << "Enter the coin denominations:" << endl;
    for (int i = 0; i < n; i++) {
        cin >> coins[i];
    }
    
    // Sort coins in descending order
    sort(coins.begin(), coins.end(), greater<int>());
    
    int amount;
    cout << "Enter the amount: ";
    cin >> amount;
    
    int count = 0;
    cout << "Coins used: ";
    for (int coin : coins) {
        while (amount >= coin) {
            amount -= coin;
            count++;
            cout << coin << " ";
        }
    }
    
    if (amount != 0) {
        cout << "\nExact amount cannot be formed with the given denominations." << endl;
    } else {
        cout << "\nMinimum number of coins needed: " << count << endl;
    }
    
    return 0;
}
